### PR TITLE
Rmk36  tedit fifth round

### PR DESCRIPTION
**Pull request by rmkaplan**
This is the long awaited, next major release candidate for Tedit.  Many changes to address many lingering issues, improve maintainabilty, robustness, and performance in some areas.   TEDIT-RELEASENOTES.TEDIT highlights some of the changes.

This is a draft, may have lingering problems--testing of screen/mouse interactions is difficult.  But please give it a try.

### CODE REVIEW

**Comment by nbriggs** [comment]
Try opening the attached Slides.tedit.txt (remove the .txt) and scrolling it.  Older TEdit scrolls OK, latest TEdit dies with an arg not non-complex number `(IQUOTIENT 0 NIL)` under SCROLLW
[Slides.tedit.txt](https://github.com/user-attachments/files/17497325/Slides.tedit.txt)

The Slides file is from https://xeroxparcarchive.computerhistory.org/phylum/ctamarin/doc/talks/.index.html

**Comment by rmkaplan** [comment]
> Try opening the attached Slides.tedit.txt (remove the .txt) and scrolling it. Older TEdit scrolls OK, latest TEdit dies with an arg not non-complex number `(IQUOTIENT 0 NIL)` under SCROLLW [Slides.tedit.txt](https://github.com/user-attachments/files/17497325/Slides.tedit.txt)
> 
> The Slides file is from https://xeroxparcarchive.computerhistory.org/phylum/ctamarin/doc/talks/.index.html

Default tab was not always initialized properly, now fixed.  File was a little bit odd: a run of tabs coded as 2-bytes (255 255 0 0 9), but that was more amusing than problematic.

**Comment by nbriggs** [comment]
After this fix, again using Slides.tedit, if the TEdit window is not as tall as the image object (e.g., shorter than the Internal Stack Frame sketch) then it is impossible to scroll past the image object - either in the scroll bar, or using the scroll wheel/trackpad.

**Comment by rmkaplan** [comment]
> After this fix, again using Slides.tedit, if the TEdit window is not as tall as the image object (e.g., shorter than the Internal Stack Frame sketch) then it is impossible to scroll past the image object - either in the scroll bar, or using the scroll wheel/trackpad.

This is somewhere on the list--I think Frank raised it quite some time ago.  The venuesysout shows a little bit more, but is goofy in a different way.

The question is whether to show a line that would be partially out of the window, either above or below.   The current strategy is to suppress the line completely, so that you don't see half of the characters on an overflow  line. That's pretty clean because most windows are much higher than the typical fonts, so a line is all or nothing.  But doesn't work well with characters in a very high font...and a high image object is treated essentially as a character in a very high font.  A line doesn't show at all if any part of its image would be outside the window.

I'm not sure what the right heuristic should be.  Turn off the filter for any line containing an object?  Turn it off for lines above a given height, whether due to a font or an object?  Suggestions?

**Comment by nbriggs** [comment]
> A line doesn't show at all if any part of its image would be outside the window.

That doesn't seem to be quite what it is doing -- as you scroll into the large sketch object it *does* display the top part of the image, but won't allow you scroll past it (you can scroll towards the beginning but not towards the end).  And when playing around with scrolling it after reshaping the window to be a little bigger (but not big enough to get the whole thing)  I got another `arg not non-complex number: NIL` -- this time in `\TEDIT.SCROLLDOWN`.  BTV! attached.
[tedit-scroll.txt](https://github.com/user-attachments/files/17510161/tedit-scroll.txt)

**Comment by nbriggs** [comment]
If it knows that the total height of a line (font, or image object) is larger than the height of the window could it not filter it out? Looking at modern apps, they don't filter at all - they jump by the height of the font you're in for input, but once the (large) text is there you can scroll to expose only a partial character.
image

The other complication is that the size of the "visible area of document" in the scroll bar keeps changing in size as you scroll:
Screenshot 2024-10-24 at 10 02 44 AM
Screenshot 2024-10-24 at 10 03 24 AM

File: SAVE.SYSOUT
Comment (nbriggs) [inline SAVE.SYSOUT:1]: Is this SAVE.SYSOUT supposed to be checked in as part of this?

**Comment by rmkaplan** [comment]
No, sorry about that.

**Comment by nbriggs** [comment]
Try this:  Open a moderate sized TEDIT window and type
```
asdf
asdf
asdf
qwer
``
leaving the caret at the end of `qwer`, and then in the left margin area (so it flips to the right slanted cursor) shift-select the `qwer`, and without moving the cursor or caret, shift-select it (the line, or where the line might have been) again and again and again.

**Comment by rmkaplan** [comment]
On the SAVE.SYSOUT:  I think the problem is that .gitignore only excludes *.sysout, not also *.SYSOUT

**Comment by rmkaplan** [comment]
(Actually, it only excludes the specific lisp.sysout, full.sysout, fuller.sysout, and apps.sysout.  Should it exclude all sysout files?)

**Comment by rmkaplan** [comment]
On the qwer, is the problem that it shows the highlighting of the insertion (maybe as a line below) but it doesn't actually display the characters?

**Comment by nbriggs** [comment]
Screen recording attached...

https://github.com/user-attachments/assets/df1c9259-b1e0-48ef-89bf-3a7af61598f4

**Comment by nbriggs** [comment]
> (Actually, it only excludes the specific lisp.sysout, full.sysout, fuller.sysout, and apps.sysout. Should it exclude all sysout files?)

It ignores the named sysouts in the loadups/ directory, but nothing at the top level.

**Comment by rmkaplan** [comment]
Give it another try.  Also look at the large image objects in the slide example.

**Comment by nbriggs** [comment]
I built with the latest changes.  Here's a screen recording of what's going on.  Notice where the mouse is when I'm doing line selections (often not within the range of the characters in the document).  Notice how sometimes it scrolls everything up off the top even though it would all fit on the screen.

https://github.com/user-attachments/assets/1a30521e-f1ed-4ba4-bae8-8d725bd44cab

**Comment by rmkaplan** [comment]
Hard to see exactly what you did, but I think I was able to see the same things.  The copy was “normalizing" the caret (early code=move to top) instead of using the new (and hopefully more intuitive) caret-scroller.  I’m trying to avoid big jumps in the caret, it’s a little tricky, there may be more.

**Comment by nbriggs** [comment]
Try scrolling backwards with smallish increments through the tall image objects in the Slides.tedit.  The scrolling feels a bit off - I'm using mouse click scrolling not scroll-wheel.  Screen recording attached.

https://github.com/user-attachments/assets/b5f4f99b-4bba-455c-9130-84dfaa29fa1d

**Comment by rmkaplan** [comment]
@nbriggs, I wasn't able to see the behavior in you last screen movie, but that may be because I continued to rework the scrolling edge cases (large objects, eol's adding and deleting in different contexts, etc.).  With the current commit, the only odd thing I have seen with that very large bit map is that if it is scrolled in just the right way, the screen is blank--and it comes back in if you jiggle it just a little bit.  So I'm still looking at that.

If you want to test scrolling a little more systematically, you can load TEDIT-DEBUG.LCOM, and then run (TSCROLL* N).  Every time you type U it will scroll the default Tedit window up by N points, typing D will scroll it down by N. Holding either key down will give repeated scrolling.  I have wheel scroll set at 20, so (TSCROLL* 20) tends to simulate a spin of the wheel.

**Comment by nbriggs** [comment]
Made a new full.sysout with latest TEdit.  Went to the end of Slides.tedit and then used `(TSCROLL* 4)` with `u` to scroll backwards -- things that work fine scrolling forwards jump around when going backwards, and certain table headings (like "Internal Stack Frame") get stuck at the top while the image object scrolls over the header leaving partial text at the top of the screen.  Also, clicking in another window and then hitting `u` generated a `CL::%%NOT-NONCOMPLEX-NUMBER`  error, but that's probably only relevant when using `TSCROLL*`.  Images attached.
Screenshot 2024-11-30 at 8 51 05 PM
Screenshot 2024-11-30 at 8 51 16 PM
Screenshot 2024-11-30 at 8 51 25 PM
Screenshot 2024-11-30 at 8 51 34 PM
Screenshot 2024-11-30 at 8 51 42 PM
Screenshot 2024-11-30 at 8 52 13 PM

**Comment by nbriggs** [comment]
Also, scrolling back-and-forth (in Slides.tedit) over "Cons Cells" and the image object below it, at the bottom of the screen, makes the image object disappear at odd times, and when you scroll back, the spacing between the title and the image changes quite noticeably.
Screenshot 2024-11-30 at 9 01 44 PM
Screenshot 2024-11-30 at 9 02 08 PM
Screenshot 2024-11-30 at 9 02 21 PM

**Comment by rmkaplan** [comment]
@nbriggs , I still haven't been able to reproduce the scrolling white-outs that you saw.  Maybe the region parameters were slightly different, in that (WINDOWPROP W 'REGION) doesn't actually return the parameter that you give to CREATEW--it doesn't account for the promptwindow.

So, if you can make it happen again, send me the value of the region parameter that you actually gave to CREATEW.

Here's a more elaborate theory of why the iterative scrolling slows down noticeably when it is dealing with the large imageobjects in the SLIDES file:

When such an image object is totally inside the window, there should be no difference in speed--the image of the imageobj is entirely within the bitmap that is blitted from one place to another, like all other lines.  (The Newman paper actually mentions using the screen bitmap in this way as another trick in Bravo).  But if the imageobj is partially off-window and then is scrolled even a little bit, the image object's displayfn has to be called to produce the part of the imageobject that was previously invisible and now comes into the window.  In this case it has to invoke all of the sketch machinery to recreate the whole image--the displayfn doesn't know how to produce just the newly exposed top or bottom.

The slowdown isn't really apparent when there is user-interaction between the scrolling events, but pretty obvious when you hold down the u or d keys in TSCROLL*.

**Comment by nbriggs** [comment]
It happens every time, for all the window sizes that I've swept out by hand, doing `(TEDIT "Slides.tedit")` -- here's a trace of the calls to `CREATEW` -- and screenshots showing the problem again

[createw.txt](https://github.com/user-attachments/files/17988534/createw.txt)

Screenshot 2024-12-02 at 10 39 38 PM
Screenshot 2024-12-02 at 10 40 07 PM
Screenshot 2024-12-02 at 10 40 21 PM
Screenshot 2024-12-02 at 10 40 31 PM

As far as the speed of scrolling goes - I don't think it's the performance, it's the amount that it actually scrolls the window for each "u" or "d" it's shifting from scrolling about 1 line per u/d to somewhere around 1-4 pixels per u/d when it's moving slowly.

**Comment by rmkaplan** [comment]
OK, I can see it. You scroll down and the object disappears, you do it again and a piece of it comes back.  Probably another case where the image is all set up, but the  clear-below-line clean-up function is called on the wrong line and wiping the image out.

The scrolling distance that you specify is really a minimum, since it is still trying to honor the constraints that it doesn't show partial lines and moves at least one line no matter what distance is specified (even 1 point, although I haven't tried that).  So it will move at least 4, but if it takes 10 points to get to the next full line, it will move 10 (or more).  But if the line is "tall" (= won't fit in the pane at all and/or is more than 50 points) and isn't already completely in the window, it gives up on the line constraints and takes the specified distance literally.  So if you specified 4, that's what you get.  And it then takes a lot more u's and d's to move it along, until it clears.

**Comment by rmkaplan** [comment]
(BTW, if you break \TEDIT.CLEARPANE.BELOW.LINE just before the "d" that gives the incorrect white space, you will see what is supposed to be showing)

**Comment by rmkaplan** [comment]
Clearing under the bitmap before printing the bottom lines seems to eliminate the last reported scrolling anomaly.

**Comment by pamoroso** [comment]
Here are some issues round 5 likely fixes but there may be others:

* #1677
* #1697
* #1738
* #1806

**Review by masinter — APPROVED** [review]
TEdit fifth round extensively tested.
12/23/2024 meeting

**Pull request merged**